bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -281,9 +449,177 @@ impl AuthenticationService {
+    /// succeed. This includes performing dynamic client registration against
+    /// the homeserver's issuer or restoring a previous registration if one has
+    /// been stored.
+    async fn configure_oidc(
+        &self,
+        oidc: &Oidc,
+        authentication_server: AuthenticationServerInfo,
+        configuration: &OidcConfiguration,
+    ) -> Result<(), AuthenticationError> {
+        if oidc.client_credentials().is_some() {
+            tracing::info!("OIDC is already configured.");
+            return Ok(());
+        };
+
+        let oidc_metadata = self.oidc_metadata(configuration)?;
+
+        if self.load_client_registration(oidc, oidc_metadata.clone()).await {
+            tracing::info!("OIDC configuration loaded from disk.");
+            return Ok(());
+        }
+
+        tracing::info!("Registering this client for OIDC.");
+        let registration_response = oidc
+            .register_client(&authentication_server.issuer, oidc_metadata.clone(), None)
+            .await?;
+
+        let client_data = RegisteredClientData {
+            // The format of the credentials changes according to the client metadata that was sent.
+            // Public clients only get a client ID.
+            credentials: ClientCredentials::None {
+                client_id: registration_response.client_id.clone(),
+            },
+            metadata: oidc_metadata,
+        };
+        oidc.restore_registered_client(authentication_server, client_data).await;
+
+        tracing::info!("Persisting OIDC registration data.");
+        self.store_client_registration(oidc).await?;
+
+        Ok(())
+    }
+
+    /// Stores the current OIDC dynamic client registration so it can be re-used
+    /// if we ever log in via the same issuer again.
+    async fn store_client_registration(&self, oidc: &Oidc) -> Result<(), AuthenticationError> {
+        let issuer = oidc.issuer().ok_or(AuthenticationError::OidcNotSupported)?;
+        let client_id = oidc
+            .client_credentials()
+            .ok_or(AuthenticationError::OidcError {
+                message: String::from("Missing client registration."),
+            })?
+            .client_id()
+            .to_owned();
+
+        let registrations = OidcRegistrations::new(
+            &self.base_path,
+            self.oidc_configuration
+                .as_ref()
+                .map(|c| c.static_registrations.clone())
+                .unwrap_or_default(),
+        )?;
+        registrations.set_client_id(client_id, issuer.to_owned())?;
 
-        Ok(HomeserverLoginDetails { url, authentication_issuer, supports_password_login })
+        Ok(())
+    }
+
+    /// Attempts to load an existing OIDC dynamic client registration for the
+    /// currently configured issuer.
+    async fn load_client_registration(
+        &self,
+        oidc: &Oidc,
+        oidc_metadata: VerifiedClientMetadata,
+    ) -> bool {
+        let Some(issuer) = oidc.issuer() else {
```

Comment:
I should have noticed that earlier, but you can't get it from the OIDC API here because we set it with `restore_registered_client`.

You should just use the value in `AuthenticationServerInfo` provided in `configure_oidc`. Same below.